enterprise/server/backends/pubsub/pubsub.go

```diff
@@ -138,8 +138,121 @@ func (p *ListPubSub) Subscribe(ctx context.Context, channelName string) interfac
 			}
 		}
 	}()
-	return &listSubscriber{
+	return &channelSubscriber{
 		cancel: cancel,
 		ch:     ch,
 	}
 }
+
+// To keep things simple for now, we maintain streams with a single value per element stored under the following key.
+const streamDataField = "data"
+
+type StreamPubSub struct {
+	rdb *redis.Client
+}
+
+// NewStreamPubSub creates a PubSub client based on a Redis-stream.
+func NewStreamPubSub(redisClient *redis.Client) *StreamPubSub {
+	return &StreamPubSub{rdb: redisClient}
+}
+
+func (p *StreamPubSub) subscribe(ctx context.Context, channelName string, startFromTail bool) interfaces.Subscriber {
+	ch := make(chan string)
+	ctx, cancel := context.WithCancel(ctx)
+
+	deliverMsg := func(msg *redis.XMessage) bool {
+		data, ok := msg.Values[streamDataField]
+		if !ok {
+			log.Errorf("Message %q on stream %q missing data field", msg.ID, channelName)
+			return false
+		}
+		str, ok := data.(string)
+		if !ok {
+			log.Errorf("Message %q on stream %q is of type %T, wanted string", msg.ID, channelName, data)
+			return false
+		}
+
+		select {
+		case ch <- str:
+			return true
+		case <-ctx.Done():
+			return false
+		}
+	}
+
+	go func() {
+		defer close(ch)
+
+		// Start from beginning of stream.
+		streamID := "0"
+
+		// If starting from tail, check if the stream has any elements.
+		// If it does then publish the last element and subscribe to messages following that element.
+		if startFromTail {
+			msgs, err := p.rdb.XRevRangeN(ctx, channelName, "+", "-", 1).Result()
+			if err != nil {
+				log.Errorf("Unable to retrieve last element of stream %q: %s", channelName, err)
+				return
+			}
+			if len(msgs) == 1 {
+				msg := msgs[0]
+				streamID = msg.ID
+				if ok := deliverMsg(&msg); !ok {
+					return
+				}
+			}
+		}
+
+		for {
+			result, err := p.rdb.XRead(ctx, &redis.XReadArgs{
+				Streams: []string{channelName, streamID},
+				Block:   0, // Block indefinitely.
+			}).Result()
+			if err != nil {
+				if err != context.Canceled {
+					log.Errorf("Error reading from stream %q: %s", channelName, err)
+				}
+				return
+			}
+
+			// We are subscribing to a single stream so there should be exactly one response.
+			if len(result) != 1 {
+				log.Errorf("Did not receive exactly one result for channel %q, got %d", channelName, len(result))
+				return
+			}
+
+			for _, msg := range result[0].Messages {
+				if ok := deliverMsg(&msg); !ok {
+					return
+				}
+			}
+		}
+	}()
+	return &channelSubscriber{
+		cancel: cancel,
+		ch:     ch,
+	}
+}
+
+// SubscribeHead returns a subscription for all previous and future message on the stream.
+func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {
+	// Subscribe from the beginning of the stream.
+	return p.subscribe(ctx, channelName, false /*startFromTail=*/)
+}
+
+// SubscribeTail returns a subscription for messages starting from the last message already on the stream, if any.
```

Comment:
ubernit, maybe we could rephrase these to make them clearer?

// SubscribeHead returns a subscription for all past, current, and future messages on the stream.
func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {}

// SubscribeTail returns a subscription for all current and future messages on the stream.
func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {}

Reply:
The term "current" is a bit ambiguous though as it depends on what the data means to the user of the client.